### ISSUE TITLE
Docker-compose, nginx, ctrl+c, gracefull stop problem

### ISSUE THREAD

**Issue by aswarcewicz** [just-containers/s6-overlay#469]
Maybe somebody could tell my what wrong with my Dockerfile.

Dockerfile (most important part):

```
ENTRYPOINT ["/init"]
CMD ["nginx", "-g", "daemon off;"]
```

then: 

1. docker-compose up
2. ctrl+c

and docker-compose wait 10s. 
It look like main process not get stop signal.

When I use dockerfile 
without `ENTRYPOINT ["/init"]`
everything is fine and nginx stop immediately

**Comment by skarnet** [comment]
`^C` is a hack and it is difficult to make it work in all cases.

When you have a CMD, under s6-overlay, you need to give the container a terminal in order for `^C` to be transmitted to the CMD. In other words, you need to run `docker run -it` instead of `docker run`. I don't know how this translates to docker-compose parlance, but there must be an option to run your container in a terminal.

**Comment by aswarcewicz** [comment]
docker-compose stop from other terminal don’t work too.

**Comment by skarnet** [comment]
I'm not sure what `docker-compose stop` does, but if it sends the equivalent of `docker stop`, then yes, it works: it sends a SIGTERM to pid 1 of the container, which stops the container. If you aren't seeing that, then your installation has other, unrelated problems.

It is obviously possible to use s6-overlay without attaching a terminal; you just won't be able to send a `^C` to your CMD. When you use s6-overlay, any signal you send is sent to the container's pid 1, not to the CMD; you control the container, not the application. This is a direct consequence of using a multiprocess system inside a container.

**Comment by skarnet** [comment]
As I _just_ said, when you send signals to a container via `docker stop`, the signal is sent to the container's pid 1, _not_ to the CMD, so it is working as intended.

However, since you're not the first person expecting the signal to be sent to the CMD instead, I will work on an option to allow that.

**Comment by skarnet** [comment]
There, I implemented support for a new variable: `S6_CMD_RECEIVE_SIGNALS`, which, when nonzero, ensures that signals from `docker stop` are rerouted from pid 1 to the CMD.

If you are able to build s6-overlay from source (it's not hard), please build the latest git head and tell me if it's working for you.

**Comment by aswarcewicz** [comment]
Now it works great, as I expected before. Thanks!

**Comment by aswarcewicz** [comment]
I forgot about S6_CMD_RECEIVE_SIGNALS variable. It works without it

**Comment by aswarcewicz** [comment]
And still problem with docker-compose maybe you now what is the difference between standard docker and docker-compose?

**Comment by skarnet** [comment]
If you want me to help you, you'll need to give more details. What exactly you are running, what your Dockerfile (or the docker-compose equivalent) is, what you expect, what you are seeing instead. The contents of your `/run/uncaught-logs/current` file if you can't see the output from your container. I can't play guessing games.

**Comment by aswarcewicz** [comment]
Dockerfile:

s6 build from source (branch master)

```
ARG NGINX_VERSION=1.21.6

FROM nginx:$NGINX_VERSION-alpine

COPY ./s6-overlay-noarch.tar.xz /tmp/s6-overlay-noarch.tar.xz
RUN tar -C / -Jxpf /tmp/s6-overlay-noarch.tar.xz
COPY ./s6-overlay-x86_64.tar.xz /tmp/s6-overlay-x86_64.tar.xz
RUN tar -C / -Jxpf /tmp/s6-overlay-x86_64.tar.xz
RUN rm -rf /tmp/*

ENV S6_KEEP_ENV=1
ENTRYPOINT ["/init"]
CMD ["nginx", "-g", "daemon off;"]
```

1. docker build -t s6test .
2. docker run --rm --name s6test s6test
3. I don't know how I test it yesterday... `ctrl+c` work but `docker stop s6test` not

**Comment by aswarcewicz** [comment]
I'm still testing for different cases but for now I have problem only with `ctrl+c` from `docker-compose up`

Dockerfile:

s6 build from source (branch master)

```
ARG NGINX_VERSION=1.21.6

FROM nginx:$NGINX_VERSION-alpine

COPY ./s6-overlay-noarch.tar.xz /tmp/s6-overlay-noarch.tar.xz
RUN tar -C / -Jxpf /tmp/s6-overlay-noarch.tar.xz
COPY ./s6-overlay-x86_64.tar.xz /tmp/s6-overlay-x86_64.tar.xz
RUN tar -C / -Jxpf /tmp/s6-overlay-x86_64.tar.xz
RUN rm -rf /tmp/*

ENV S6_KEEP_ENV=1
ENV S6_CMD_RECEIVE_SIGNALS=1
ENTRYPOINT ["/init"]
CMD ["nginx", "-g", "daemon off;"]
```

1. `docker build -t s6test .`
2. `docker run --rm --name s6test s6test`
3. `ctrl+c` work and `docker stop s6test` too

docker-compose file:
```
version: '3'

services:
  nginx:
    build:  .
    container_name: s6composetest
    ports:
      - 8081:80
```

1. docker-compose build
2. docker-compose up
3. `ctrl+c` and here is problem. No signal passed

**Comment by skarnet** [comment]
Yeah so this is `docker-compose` not having the same UI as `docker run`, and has nothing to do with s6-overlay.

When you run `docker run` (without `-it`), and you ^C it, it specialcases it and sends a signal (I think it's SIGINT, I could be wrong though) to the container's pid 1. That is what allows you to stop the CMD or the container by interrupting `docker run` even though the container isn't running in a terminal.

When you run `docker-compose`, apparently, it does not have that special case, and when you ^C it, it *does not* send any signal to the container. There may be a configuration option to `docker-compose` to do that, but my suspicion is that it's simply not meant to be used interactively this way.

In other words: **stop assuming ^C is a valid way to stop a container. It's not.**

**Comment by socheatsok78** [comment]
You can try this approach.

Create a new file `/docker-cmd.sh`, and override the `nginx` official image `CMD [ "/docker-cmd.sh" ]`.

```sh
#!/bin/sh

# trap SIGINT
trap '/run/s6/basedir/bin/halt' SIGINT

# noop
while true; do sleep 30; done;
```

Then start your `nginx -g "daemon off;"` as part of `s6-rc.d` service.

**Comment by socheatsok78** [comment]
I've made an example here, https://github.com/socheatsok78/s6-overlay-nginx-module.   
I'm not recommend you to use this, but feel free to make a copy and try it out yourself.

**Comment by aswarcewicz** [comment]
ok I run some tests and now everything looks fine.
Thank you for your work!
